pkg/packing/packer.go

```diff
@@ -0,0 +1,151 @@
+	for len(remainingPods) > 0 {
+		packing, remainingPods = p.packWithLargestPod(remainingPods, instances, constraints)
+		// checked all instance types and found no packing option
+		if len(packing.Pods) == 0 {
+			zap.S().Warnf("Failed to compute packing for pod(s) %v with instance type option(s) %v", apiobject.PodNamespacedNames(remainingPods), instanceTypeNames(instances))
+			remainingPods = remainingPods[1:]
+			continue
+		}
+		packings = append(packings, packing)
+		zap.S().Debugf("Successfully computed packing for pod(s) %v with instance type option(s) %s", apiobject.PodNamespacedNames(packing.Pods), instanceTypeNames(packing.InstanceTypeOptions))
+	}
+	return packings
+}
+
+// packWithLargestPod will try to pack max number of pods with largest pod in
+// pods across all available node capacities. It returns Packing: max pod count
+// that fit; with their node capacities and list of leftover pods
+func (p *packer) packWithLargestPod(unpackedPods []*v1.Pod, instances []cloudprovider.InstanceType, constraints *cloudprovider.Constraints) (*Packing, []*v1.Pod) {
+	bestPackedPods := []*v1.Pod{}
+	bestInstances := []cloudprovider.InstanceType{}
+	remainingPods := unpackedPods
+	for _, packable := range PackablesFrom(instances, constraints.Overhead) {
+		// check how many pods we can fit with the available capacity
+		result := packable.Pack(unpackedPods)
+		if len(result.packed) == 0 {
+			continue
+		}
+		// If the pods packed are the same as before, this instance type can be
+		// considered as a backup option in case we get ICE
+		if p.podsMatch(bestPackedPods, result.packed) {
+			bestInstances = append(bestInstances, packable.InstanceType)
+		} else if len(result.packed) > len(bestPackedPods) {
+			// If pods packed are more than compared to what we got in last
+			// iteration, consider using this instance type
+			bestPackedPods = result.packed
+			remainingPods = result.unpacked
+			bestInstances = []cloudprovider.InstanceType{packable.InstanceType}
+		}
+	}
+	sortByResources(bestInstances)
+	return &Packing{Pods: bestPackedPods, InstanceTypeOptions: bestInstances}, remainingPods
+}
+
+func (*packer) podsMatch(first, second []*v1.Pod) bool {
+	if len(first) != len(second) {
+		return false
+	}
+	podSeen := map[string]int{}
+	for _, pod := range first {
+		podSeen[apiobject.NamespacedName(pod).String()]++
+	}
+	for _, pod := range second {
+		podSeen[apiobject.NamespacedName(pod).String()]--
+	}
+	for _, value := range podSeen {
+		if value != 0 {
+			return false
+		}
+	}
+	return true
+}
+
+// sortByResources sorts instance type packings by vcpus and memory resources
+func sortByResources(instances []cloudprovider.InstanceType) {
+	sort.Slice(instances, func(i, j int) bool { return weightOf(instances[i]) < weightOf(instances[j]) })
+}
+
+func weightOf(instance cloudprovider.InstanceType) float64 {
+	return euclidean(
+		float64(instance.CPU().Value()),
+		float64(instance.Memory().ScaledValue(resource.Mega)), // 1 gb = 1 cpu
```

Comment:
With this we will always prefer instance with higher memory when CPU is the same, for example -`t2.large 2cpu-8gb` over `t2.medium 2cpu-4gb`?

Reply:
This doesn't change the current logic at all, it just refactors it. The sort uses a `<` operator, so IIUC it will prefer smaller.